compiler/rustc_resolve/src/late/lifetimes.rs

```diff
@@ -1318,13 +1318,49 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
                         && let hir::IsAsync::NotAsync = self.tcx.asyncness(lifetime_ref.hir_id.owner.def_id)
                         && !self.tcx.features().anonymous_lifetime_in_impl_trait
                     {
-                        rustc_session::parse::feature_err(
-                            &self.tcx.sess.parse_sess,
-                            sym::anonymous_lifetime_in_impl_trait,
-                            lifetime_ref.span,
-                            "anonymous lifetimes in `impl Trait` are unstable",
-                        ).emit();
-                        return;
+
+                        match self.tcx.hir().get_generics(lifetime_ref.hir_id.owner.def_id) {
+                            Some(generics) => {
+                                for i in 0..generics.params.len()  {
+
+                                    if !generics.span.contains(generics.params[i].span)  {
+
+                                        let mut diag =  rustc_session::parse::feature_err(
+                                            &self.tcx.sess.parse_sess,
+                                            sym::anonymous_lifetime_in_impl_trait,
+                                            lifetime_ref.span,
+                                            "anonymous lifetimes in `impl Trait` are unstable",
+                                        );
+
+                                        diag.span_label(lifetime_ref.span, "expected named lifetime parameter");
+                                        diag.multipart_suggestion("consider introducing a named lifetime parameter",
+                                                                  vec![
+                                                                    (lifetime_ref.span.shrink_to_hi(), "'a ".to_owned()),
+                                                                    match generics.span_for_param_suggestion() {
+                                                                        Some(_) => {
+                                                                            (self.tcx.sess.source_map().span_through_char(generics.span, '<').shrink_to_hi(), "'a, ".to_owned())
+                                                                        }
+                                                                        None => (generics.span, "<'a>".to_owned()),
+
+                                                                    }
+                                                                  ], rustc_errors::Applicability::MaybeIncorrect);
+                                        diag.emit();
+
+                                        return;
+                                    }
+                                }
```

Comment:
What happens if we leave the loop without emitting the diagnostic?  For instance if we have `fn foo<>()`.